x/stake/keeper/slash.go

```diff
@@ -4,56 +4,212 @@ import (
 	validator, found := k.GetValidatorByPubKey(ctx, pubkey)
 	if !found {
-		panic(fmt.Errorf("Attempted to slash a nonexistent validator with address %s", pubkey.Address()))
+		panic(fmt.Errorf("attempted to slash a nonexistent validator with address %s", pubkey.Address()))
 	}
-	sharesToRemove := validator.PoolShares.Amount.Mul(fraction)
+	ownerAddress := validator.GetOwner()
+
+	// Track remaining slash amount
+	remainingSlashAmount := slashAmount
+
+	switch {
+	case infractionHeight > ctx.BlockHeight():
+		// Can't slash infractions in the future
+		panic(fmt.Sprintf("impossible attempt to slash future infraction at height %d but we are at height %d", infractionHeight, ctx.BlockHeight()))
+
+	case infractionHeight == ctx.BlockHeight():
+		// Special-case slash at current height for efficiency - we don't need to look through unbonding delegations or redelegations
+		logger.Info(fmt.Sprintf("Slashing at current height %d, not scanning unbonding delegations & redelegations", infractionHeight))
+
+	case infractionHeight < ctx.BlockHeight():
+		// Iterate through unbonding delegations from slashed validator
+		unbondingDelegations := k.GetUnbondingDelegationsFromValidator(ctx, ownerAddress)
+		for _, unbondingDelegation := range unbondingDelegations {
+			amountSlashed := k.slashUnbondingDelegation(ctx, unbondingDelegation, infractionHeight, fraction)
+			if amountSlashed.IsZero() {
+				continue
+			}
+			remainingSlashAmount = remainingSlashAmount.Sub(amountSlashed)
+		}
+
+		// Iterate through redelegations from slashed validator
+		redelegations := k.GetRedelegationsFromValidator(ctx, ownerAddress)
+		for _, redelegation := range redelegations {
+			amountSlashed := k.slashRedelegation(ctx, validator, redelegation, infractionHeight, fraction)
+			if amountSlashed.IsZero() {
+				continue
+			}
+			remainingSlashAmount = remainingSlashAmount.Sub(amountSlashed)
+		}
+
+	}
+
+	// Cannot decrease balance below zero
+	sharesToRemove := remainingSlashAmount
+	if sharesToRemove.GT(validator.PoolShares.Amount.EvaluateInt()) {
+		sharesToRemove = validator.PoolShares.Amount.EvaluateInt()
+	}
+
+	// Get the current pool
 	pool := k.GetPool(ctx)
-	validator, pool, burned := validator.RemovePoolShares(pool, sharesToRemove)
-	k.SetPool(ctx, pool)              // update the pool
-	k.UpdateValidator(ctx, validator) // update the validator, possibly kicking it out
+	// remove shares from the validator
+	validator, pool, burned := validator.RemovePoolShares(pool, sdk.NewRatFromInt(sharesToRemove))
+	// burn tokens
+	pool.LooseTokens -= burned
+	// update the pool
+	k.SetPool(ctx, pool)
+	// update the validator, possibly kicking it out
+	k.UpdateValidator(ctx, validator)
 
-	logger := ctx.Logger().With("module", "x/stake")
+	// Log that a slash occurred!
 	logger.Info(fmt.Sprintf("Validator %s slashed by fraction %v, removed %v shares and burned %d tokens", pubkey.Address(), fraction, sharesToRemove, burned))
+
```

Comment:
Can there be a sanity check that that the remaining slash amount = 0?

Reply:
It's not necessarily zero; most of the stake may still be bonded.